### PR TITLE
Feature/ast import statements

### PR DESCRIPTION
**Pull request by brewsoftware**
Added syntax to AST for including import {'module'} decelerations.

I consider this as Level 0 - Open for comment and please pull this request if your happy to accept it for later.

Level 1  -> Include better path validation
Level 2 -> Tag as local or global modules (beginning with a "./" or without indicates local/global)

Cheers,
Jason

### CODE REVIEW

**Comment by brewsoftware** [comment]
Hey Guys,
   So I've re-written the compiler part of this project as a gulp plugin with support for import statements. It can be found here https://github.com/brewsoftware/gulp-firebase-bolt/ 

I have a few issues to fix on this PR relating to the parser but will have an updated version shortly. Please feel free to check it out and let me know if there any issues.

Cheers,
Jason

**Comment by brewsoftware** [comment]
So I've fixed the PR issue that was blocking this earlier but the issue is coming from master. Happy for you to merge this but please indicate if you want to review the symtax. It appears to be working fine with some of the downstream testing that I'm doing. When this is merged we can maybe update the core compiler as well to include the concurrent file update.

File: src/rules-generator.ts [hunk 1]
```diff
@@ -652,6 +652,12 @@ export class Generator {
       return params[exp2.name] || self.globals[exp2.name] || exp2;
     }
 
+    // Convert ref[prop] => ref.child(prop)
```
Comment (brewsoftware) [inline src/rules-generator.ts]: This is just hoisted from below as it's causing some of the builds to fail.

**Comment by mckoss** [comment]
Thanks for the contribution - but I don't think I can merge this as is.   I'm not sure I agree that

```
import {'filename'}
```

is the right syntax here - it diverges from ES6 syntax, so would need good justification to do so.

Might it make more sense to use:

```
import * as 'externalModule' from './module';
import {'CustomType1', 'CustomType2'} from './module';
```

As an aside, from my quick scan of your code - I see some non-standard formatting as well - surprised it was not caught by tslint.

File: src/ast.ts [hunk 2]
```diff
@@ -99,6 +99,11 @@ export interface Method {
   body: Exp;
 }
 
+export interface Import {
+  filename: string;
+  alias: string;
+  scope: boolean;
+}
 export class PathPart {
```
Comment (mckoss) [inline src/ast.ts]: Missing blank line.

File: src/bolt.ts [hunk 3]
```diff
@@ -20,7 +20,6 @@ import generator = require('./rules-generator');
 import simulator = require('./simulator');
 import astReal = require('./ast');
 import util = require('./util');
-
```
Comment (mckoss) [inline src/bolt.ts]: Why did you remove this spacing between imports and exports?

File: src/rules-generator.ts (hunk 1 above)
Comment (mckoss) [inline src/rules-generator.ts]: I merged this change in yesterday.

File: src/rules-parser.pegjs [hunk 4]
```diff
@@ -84,7 +84,26 @@ start = _ Statements _ {
 
 Statements = rules:(Statement _)*
 
-Statement = f:Function / p:Path / s:Schema
+Statement = f:Function / p:Path / s:Schema / i:Import
+AliasName "aliasname definition" = " as " _ name:Identifier{
```
Comment (mckoss) [inline src/rules-parser.pegjs]: Separate with blank line.

File: src/rules-parser.pegjs [hunk 5]
```diff
@@ -84,7 +84,26 @@ start = _ Statements _ {
...
+  return name;
+}
+Import "import definition" = body:"import {'"_ scope:("./" / "../")? _  value:FilePath _"'}"  alias:(AliasName)? {
```
Comment (mckoss) [inline src/rules-parser.pegjs]: missing blank line

File: src/rules-parser.pegjs [hunk 6]
```diff
@@ -84,7 +84,26 @@ start = _ Statements _ {
...
+  var rebuildValue = value;
+  if(scope){
+    rebuildValue = scope + rebuildValue;
+  }
+  var an = "";
+  if(alias){
+    an = alias;
+  }
+  symbols.registerImport(alias, rebuildValue, scope);
+
```
Comment (mckoss) [inline src/rules-parser.pegjs]: remove blank line

**Comment by mckoss** [comment]
Is this an incomplete CL?  Seems like it adds syntax, but does not actually perform an import of an external Bolt file.

**Comment by brewsoftware** [comment]
Hey, thanks for the feedback. Yes your observations are correct. Specifically the debate around aliasing and es6 syntax was what i was looking for.

The original intention was to nut out the syntax seperatly to the compiler. Linting updates would be good if theres an issue there...

I do have a compiler that works with the ast but its written in gulp/through2. I can port this back over next week but it requires a rather large change to the file handling/async processing of the bolt compiler that could be considered a breaking change...

**Comment by brewsoftware** [comment]
So import statements. The question i have is 
1. Importing types/functions is straight forward. Aliasing should work. Recursive aliasing may become a bit of an mess to build in the ast.
2. Import paths? I'm stuck on this one (or more specifically should we support this at this time)
3. Should we be explicit with the different import behaviours and have 2/3 import syntaxes?

**Comment by mckoss** [comment]
I'd like to close this PR as we won't be merging as is.  I prefer to move discussion to the Issue #63
